docs/src/diplomacy/adder_tutorial.md

```diff
@@ -0,0 +1,322 @@
+> Diplomacy is a parameter negotiation framework for generating parameterized 
+  protocol implementations.
+
+The goal of this walkthrough is to demonstrate an extremely simple Diplomacy
+protocol. In this walkthrough, we will demonstrate how to create a parameterized
+adder and its associated testing modules.
+
+## Diplomatic Adder and Test Bench
+
+Let's begin by describing the desired circuit. We want a 2-to-1 **adder**, and
+to test it, we would like a **top-level test bench**, a **driver**, and a simple
+**monitor**.
+
+![adder test bench module hierarchy](diagrams/adder_modules.pdf)
+
+Let's say we would also like to parameterize the widths of these modules' ports.
+For the sake of demonstration, when negotiating parameters, our protocol will
+use the smaller of the two widths, which is the opposite of what typical Chisel
+width inference would do.
+
+```scala mdoc:invisible
+import chipsalliance.rocketchip.config.{Config, Parameters}
+import chisel3._
+import chisel3.internal.sourceinfo.SourceInfo
+import chisel3.stage.ChiselStage
+import chisel3.util.random.FibonacciLFSR
+import freechips.rocketchip.diplomacy.{SimpleNodeImp, RenderedEdge, ValName, SourceNode,
+                                       NexusNode, SinkNode, LazyModule, LazyModuleImp}
+```
+
+### Parameter negotiation and passing
+
+We have a width parameter we would like our modules to share. Any module that
+wants to send or receive parameter information must have **one or more nodes**.
+
+Between each pair of nodes, there may be **one or more edges**. The edge does
+the negotiation work to ensure parameter agreement between the nodes it connects.
+
+Let's pass around our parameters like so:
+
+![adder diplomatic nodes](diagrams/adder_nodes.pdf)
+
+Because the graph diplomacy creates is **acyclic**, it is directional. The
+direction towards the sinks is **downward**, and the direction towards the
+sources is **upward**.
+
+#### Parameters
+
+In our example, we want all of our circuitry to agree upon a Scala `Int` to use
+as a wire width. For clarity, we will define distinct case classes for our
+parameters, although they all contain the same type.
+
+```scala mdoc
+case class UpwardParam(width: Int)
+case class DownwardParam(width: Int)
+case class EdgeParam(width: Int)
+```
+
+#### Node Implementation
+
+In our "node implementation", or `NodeImp`, we describe how parameters flow
+throughout our graph and how parameters are negotiated between nodes.
+
+We will use `SimpleNodeImp`, which performs the same parameter negotiation and
+passes the same bundles along an edge, regardless of whether the edge points
+into our out of the node.
+
+Our **edge parameter** (`E`) describes the data type that needs to be passed along
+the edges in our graph; in our case we only need to keep track of one `Int` that
+describes our width, which will live inside our `EdgeParam` class.
```

Comment:
"describes our width" => "describes the final, negotiated width used to actually create the hardware wire"